runtime/middlewares.go

```diff
@@ -0,0 +1,196 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zanzibar
+
+import (
+	"context"
+	"net/http"
+)
+
+// Handler is an interface that compatable middlewares must implement to be used
+// in the middleware stack. Handle should yield to the next middleware in the stack
+// by invoking the next HandlerFn passed in.
+type Handler interface {
```

Comment:
It can still work, just not using `defer` itself.

If a middleware's `HandleRequest()` is called then the middlewares `HandleResponse()` will be called. You can store some state on the middleware handle instance in `HandleRequest()` and explicitly clean it up in `HandleResponse()`

As for panic safety, if something panics then the process is in an undefined state and we have to restart it, I don't know if we can have panic safety for all code in the codebase.

I'm not familiar enough to talk about context scoping but it sounds like the middlewares and endpoints will all have the same context managed by the router/server itself.